data-access/UserRepository.cs

```diff
@@ -44,5 +73,262 @@ public async Task<UserEditionPermissions> GetUserEditionPermissionsAsync(uint us
+                        var tokenEntry = await connection.ExecuteAsync(CreateUserEmailTokenQuery.GetQuery, new
+                        {
+                            Token = token,
+                            UserId = userInfo.UserId,
+                            Type = CreateUserEmailTokenQuery.ResetPassword
+                        });
+                        if (tokenEntry != 1)
+                            return null;
+                        
+                        // Cleanup
+                        transactionScope.Complete();
+                        
+                        // Pass the token back in the user info object
+                        userInfo.Token = token;
+                        return userInfo;
+                    }
+                    catch{} // Suppress errors here. We don't want to risk people fishing for valid email addresses,
+                            // though any errors are suppressed in the controller too.
+                }
+            }
+            return null;
+        }
+        
+        /// <summary>
+        /// Resets a user's password based on the secret reset password token sent to the user's email address
+        /// </summary>
+        /// <param name="token">Secret token for resetting password</param>
+        /// <param name="password">New password for the user's account</param>
+        /// <returns></returns>
+        /// <exception cref="DbFailedWrite"></exception>
+        public async Task ResetForgottenPasswordAsync(string token, string password)
+        {
+            await CleanOldTokens(); // Always flush the expired tokens from the database first
+            using (var transactionScope = new TransactionScope())
+            {
+                using (var connection = OpenConnection())
+                {
+                    var resetPassword = await connection.ExecuteAsync(UpdatePasswordByToken.GetQuery, new
+                    {
+                        Token = token,
+                        Password = password
+                    });
+                    if (resetPassword != 1)
+                        throw new DbFailedWrite();
+                    var cleanTokenTable = await connection.ExecuteAsync(DeleteUserEmailTokenQuery.GetTokenQuery, new
+                    {
+                        Token = token
+                    });
+                    if (cleanTokenTable != 1)
+                        throw new DbFailedWrite();
+                    transactionScope.Complete(); // Close the transaction
+                }
+            }
+        }
+
+        /// <summary>
+        /// Clears out old tokens, but see the comments to DeleteOldTokens.
+        /// </summary>
+        /// <returns></returns>
+        private async Task CleanOldTokens()
+        {
+            using (var connection = OpenConnection())
+            {
+                await connection.ExecuteAsync(DeleteOldTokens.GetQuery, new { Days = _tokenValidDays });
```

Comment:
I see now that tokens do have expiry dates, but I couldn't see them in the queries.